File: hyper-collate-core/src/main/java/nl/knaw/huygens/hypercollate/collator/HyperCollator.java
package nl.knaw.huygens.hypercollate.collator;

/*-
 * #%L
 * HyperCollate core
 * =======
 * Copyright (C) 2017 - 2018 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import eu.interedition.collatex.Token;
import eu.interedition.collatex.dekker.Tuple;
import nl.knaw.huygens.hypercollate.model.*;
import nl.knaw.huygens.hypercollate.tools.CollationGraphRanking;
import nl.knaw.huygens.hypercollate.tools.CollationGraphVisualizer;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.*;
import java.util.function.BiFunction;
import java.util.function.Predicate;

import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;
import static nl.knaw.huygens.hypercollate.tools.StreamUtil.stream;

public class HyperCollator {
  private static final Logger LOG = LoggerFactory.getLogger(HyperCollator.class);

  private static final BiFunction<SimpleTokenVertex, SimpleTokenVertex, Boolean> matcher = //
      (stv1, stv2) -> {
        if (stv1.getNormalizedContent().isEmpty() && stv2.getNormalizedContent().isEmpty()) {
          if (stv1.getContent().isEmpty() && stv2.getContent().isEmpty()) {
            // both are milestones, so compare their tags.
            String parentTag1 = stv1.getParentXPath().replace("/.*/", "");
            String parentTag2 = stv2.getParentXPath().replace("/.*/", "");
            return parentTag1.equals(parentTag2);
          } else {
            return stv1.getContent().equals(stv2.getContent());
          }
        }
        return stv1.getNormalizedContent().equals(stv2.getNormalizedContent());
      };

  public CollationGraph collate(VariantWitnessGraph... graphs) {
    List<String> sigils = new ArrayList<>();
    List<VariantWitnessGraph> witnesses = new ArrayList<>();
    List<VariantWitnessGraphRanking> rankings = new ArrayList<>();
    Arrays.stream(graphs)//
        .sorted(comparing(VariantWitnessGraph::getSigil))//
        .forEach(graph -> {
          sigils.add(graph.getSigil());
          witnesses.add(graph);
          rankings.add(VariantWitnessGraphRanking.of(graph));
        });

    Set<Match> matches = getPotentialMatches(witnesses, rankings);
    CollationGraph collationGraph = new CollationGraph();
    Map<TokenVertex, TextNode> collatedTokenVertexMap = new HashMap<>();
    VariantWitnessGraph first = witnesses.remove(0);

    Map<String, List<Match>> matchesSortedByRankPerWitness = sortAndFilterMatchesByWitness(matches, sigils);
    Map<Markup, MarkupNode> markupNodeIndex = new HashMap<>();

    initialize(collationGraph, collatedTokenVertexMap, markupNodeIndex, first);
    visualize(collationGraph);

    for (VariantWitnessGraph witnessGraph : witnesses) {
      List<Match> sortedMatchesForWitness = matchesSortedByRankPerWitness.get(witnessGraph.getSigil());
      collate(collationGraph, //
          witnessGraph, //
          sortedMatchesForWitness, //
          markupNodeIndex, //
          collatedTokenVertexMap);
      visualize(collationGraph);
    }
    return collationGraph;
  }

  private void visualize(CollationGraph collationGraph) {
    String dot = CollationGraphVisualizer.toDot(collationGraph, true);
    System.out.println(dot);
  }

  void initialize(CollationGraph collationGraph, //
                  Map<TokenVertex, TextNode> collatedTokenVertexMap, //
                  Map<Markup, MarkupNode> markupNodeIndex,//
                  VariantWitnessGraph witnessGraph) {
    String sigil = witnessGraph.getSigil();
    collationGraph.getSigils().add(sigil);
    addMarkupNodes(collationGraph, markupNodeIndex, witnessGraph);

    collatedTokenVertexMap.put(witnessGraph.getStartTokenVertex(), collationGraph.getTextStartNode());
    stream(witnessGraph.vertices())//
        .filter(SimpleTokenVertex.class::isInstance)//
        .forEach(tokenVertex ->
            addCollationNode(collationGraph, collatedTokenVertexMap, tokenVertex, witnessGraph, markupNodeIndex)
        );
    collatedTokenVertexMap.put(witnessGraph.getEndTokenVertex(), collationGraph.getTextEndNode());
    addEdges(collationGraph, collatedTokenVertexMap);
  }

  private void addMarkupNodes(CollationGraph collationGraph, Map<Markup, MarkupNode> markupNodeIndex, VariantWitnessGraph witnessGraph) {
    witnessGraph.getMarkupStream().forEach(markup -> {
      MarkupNode markupNode = collationGraph.addMarkupNode(witnessGraph.getSigil(), markup);
      markupNodeIndex.put(markup, markupNode);
    });
  }

  private List<CollatedMatch> getCollatedMatches(Map<TokenVertex, TextNode> collatedTokenVertexMap,//
                                                 List<Match> matches, String sigil) {
    return matches.stream()//
//        .peek(System.out::println)//
        .map(match -> toCollatedMatch(match, sigil, collatedTokenVertexMap))//
        .collect(toList());
  }

  private CollatedMatch toCollatedMatch(Match match, String sigil, Map<TokenVertex, TextNode> collatedTokenVertexMap) {
    TokenVertex vertex = stream(match.getTokenVertexList())//
        .filter(tv -> !tv.getSigil().equals(sigil))//
        .findFirst()//
        .orElseThrow(() -> new RuntimeException("No vertex!"));
    TokenVertex tokenVertexForWitness = match.getTokenVertexForWitness(sigil);
    TextNode node = collatedTokenVertexMap.get(vertex);
    return new CollatedMatch(node, tokenVertexForWitness).setVertexRank(match.getRankForWitness(sigil));
  }

  private void collate(CollationGraph collationGraph, //
                       VariantWitnessGraph witnessGraph, //
                       List<Match> sortedMatchesForWitness, //
                       Map<Markup, MarkupNode> markupNodeIndex,//
                       Map<TokenVertex, TextNode> collatedTokenVertexMap) {
    Predicate<? super Match> matchesWithCollationGraph = m -> collationGraph.getSigils()//
        .stream()//
        .anyMatch(m::hasWitness);
    CollationGraphRanking baseRanking = CollationGraphRanking.of(collationGraph);

    List<Match> filteredSortedMatchesForWitness = sortedMatchesForWitness.stream()//
        .filter(matchesWithCollationGraph)//
        .collect(toList());

    String witnessSigil = witnessGraph.getSigil();
    collationGraph.getSigils().add(witnessSigil);
    addMarkupNodes(collationGraph, markupNodeIndex, witnessGraph);

    List<CollatedMatch> matchList = getCollatedMatches(collatedTokenVertexMap, filteredSortedMatchesForWitness, witnessSigil)//
        .stream()//
//        .peek(System.out::println)//
        .map(m -> adjustRankForCollatedNode(m, baseRanking))//
        .distinct()//
        .collect(toList());
    LOG.info("matchList={}", matchList);
    OptimalCollatedMatchListAlgorithm optimalCollatedMatchListAlgorithm = new OptimalCollatedMatchListAlgorithm();
    List<CollatedMatch> optimalMatchList = optimalCollatedMatchListAlgorithm.getOptimalCollatedMatchList(matchList);
//    DecisionTreeNode decisionTreeRootNode = optimalCollatedMatchListAlgorithm.getDecisionTreeRootNode();
//    visualize(decisionTreeRootNode);

    LOG.info("optimalMatchList={}", optimalMatchList);

    Iterator<TokenVertex> witnessIterator = VariantWitnessGraphTraversal.of(witnessGraph).iterator();
    TokenVertex first = witnessIterator.next();
    TextNode rootNode = collationGraph.getTextStartNode();
    collatedTokenVertexMap.put(first, rootNode);

    logCollated(collatedTokenVertexMap);
    while (!optimalMatchList.isEmpty()) {
      CollatedMatch match = optimalMatchList.remove(0);
      System.out.println(match);
      TokenVertex tokenVertexForWitnessGraph = match.getWitnessVertex();
      advanceWitness(collationGraph, collatedTokenVertexMap, witnessIterator, tokenVertexForWitnessGraph, witnessGraph, markupNodeIndex);

      TextNode matchingNode = match.getCollatedNode();
      Token token = tokenVertexForWitnessGraph.getToken();
      if (token != null) {
        matchingNode.addToken(token);
        matchingNode.addBranchPath(tokenVertexForWitnessGraph.getSigil(), tokenVertexForWitnessGraph.getBranchPath());
      }
      collatedTokenVertexMap.put(tokenVertexForWitnessGraph, matchingNode);
      addMarkupHyperEdges(collationGraph, witnessGraph, markupNodeIndex, tokenVertexForWitnessGraph, matchingNode);
    }
    collatedTokenVertexMap.put(witnessGraph.getEndTokenVertex(), collationGraph.getTextEndNode());

    logCollated(collatedTokenVertexMap);
    addEdges(collationGraph, collatedTokenVertexMap);
    logCollated(collatedTokenVertexMap);
  }

  private void addMarkupHyperEdges(CollationGraph collationGraph, VariantWitnessGraph witnessGraph, Map<Markup, MarkupNode> markupNodeIndex, TokenVertex tokenVertexForWitnessGraph, TextNode matchingNode) {
    witnessGraph.getMarkupListForTokenVertex(tokenVertexForWitnessGraph)//
        .forEach(markup -> collationGraph.linkMarkupToText(markupNodeIndex.get(markup), matchingNode));
  }

  private List<CollatedMatch> getOptimalMatchList(List<CollatedMatch> matchList) {
    return new OptimalCollatedMatchListAlgorithm().getOptimalCollatedMatchList(matchList);
  }

  private CollatedMatch adjustRankForCollatedNode(CollatedMatch m, CollationGraphRanking baseRanking) {
    TextNode node = m.getCollatedNode();
    Integer rank = baseRanking.apply(node);
    m.setNodeRank(rank);
    return m;
  }

  private void logCollated(Map<TokenVertex, TextNode> collatedTokenVertexMap) {
    List<String> lines = new ArrayList<>();
    collatedTokenVertexMap.forEach((k, v) -> lines.add(k.getSigil() + ":" + k.getToken() + " -> " + v));
    LOG.info("collated={}", lines.stream().sorted().collect(joining("\n")));
  }

  private void advanceWitness(CollationGraph collationGraph, //
                              Map<TokenVertex, TextNode> collatedTokenVertexMap, //
                              Iterator<TokenVertex> tokenVertexIterator, //
                              TokenVertex tokenVertexForWitness,//
                              VariantWitnessGraph witnessGraph,//
                              Map<Markup, MarkupNode> markupNodeIndex) {
    if (tokenVertexIterator.hasNext()) {
      TokenVertex nextWitnessVertex = tokenVertexIterator.next();
      while (tokenVertexIterator.hasNext() && !nextWitnessVertex.equals(tokenVertexForWitness)) {
        addCollationNode(collationGraph, collatedTokenVertexMap, nextWitnessVertex, witnessGraph, markupNodeIndex);
        nextWitnessVertex = tokenVertexIterator.next();
      }
    }
  }

  private static void addEdges(CollationGraph collationGraph, Map<TokenVertex, TextNode> collatedTokenVertexMap) {
    collatedTokenVertexMap.keySet()
        .forEach(tv -> tv.getIncomingTokenVertexStream()
            .forEach(itv -> {
              TextNode source = collatedTokenVertexMap.get(itv);
              TextNode target = collatedTokenVertexMap.get(tv);
              if (source == null) {
                throw new RuntimeException("source=null, target=" + target);
              }
              if (target == null) {
                throw new RuntimeException("target=null, source=" + source);
              }
              List<TextNode> existingTargetNodes = collationGraph.getOutgoingTextEdgeStream(source)
                  .map(collationGraph::getTarget)//
                  .map(TextNode.class::cast)//
                  .collect(toList());
              String sigil = tv.getSigil();
              if (!existingTargetNodes.contains(target)) {
                Set<String> sigils = new HashSet<>();
                sigils.add(sigil);
                collationGraph.addDirectedEdge(source, target, sigils);
                // System.out.println("> " + source + " -> " + target);
              } else {
                TextEdge edge = collationGraph.getOutgoingTextEdgeStream(source)
                    .filter(e -> target.equals(collationGraph.getTarget(e)))//
                    .findFirst()//
                    .orElseThrow(() -> new RuntimeException("There should be an edge!"));
                edge.addSigil(sigil);
                // System.err.println("duplicate edge: " + source + " -> " + target);
              }
            }));
  }

  private void addCollationNode(CollationGraph collationGraph, //
                                Map<TokenVertex, TextNode> collatedTokenVertexMap, //
                                TokenVertex tokenVertex,//
                                VariantWitnessGraph witnessGraph,//
                                Map<Markup, MarkupNode> markupNodeIndex) {
    if (!collatedTokenVertexMap.containsKey(tokenVertex)) {
      TextNode collationNode = collationGraph.addTextNodeWithTokens(tokenVertex.getToken());
      collationNode.addBranchPath(tokenVertex.getSigil(), tokenVertex.getBranchPath());
      collatedTokenVertexMap.put(tokenVertex, collationNode);
      addMarkupHyperEdges(collationGraph, witnessGraph, markupNodeIndex, tokenVertex, collationNode);
    }
  }

  Map<String, List<Match>> sortAndFilterMatchesByWitness(Set<Match> matches, List<String> sigils) {
    Map<String, List<Match>> map = new HashMap<>();
    sigils.forEach(s -> {
      List<Match> sortedMatchesForWitness = filterAndSortMatchesForWitness(matches, s);
      map.put(s, sortedMatchesForWitness);
    });
    return map;
  }

  private List<Match> filterAndSortMatchesForWitness(Set<Match> matches, String sigil) {
    Comparator<Match> comparator = (match1, match2) -> {
      Integer rank1 = match1.getRankForWitness(sigil);
      Integer rank2 = match2.getRankForWitness(sigil);
      if (rank1.equals(rank2)) {
        rank1 = match1.getLowestRankForWitnessesOtherThan(sigil);
        rank2 = match2.getLowestRankForWitnessesOtherThan(sigil);
      }
      return rank1.compareTo(rank2);
    };
    return matches.stream()//
        .filter(m -> m.hasWitness(sigil))//
        .sorted(comparator)//
        .collect(toList());
  }

  Set<Match> getPotentialMatches(List<VariantWitnessGraph> witnesses, List<VariantWitnessGraphRanking> rankings) {
    Set<Match> allPotentialMatches = new HashSet<>();
    Map<TokenVertex, Match> vertexToMatch = new HashMap<>();
    for (Tuple<Integer> t : permute(witnesses.size())) {
      VariantWitnessGraph witness1 = witnesses.get(t.left);
      VariantWitnessGraphRanking ranking1 = rankings.get(t.left);
      VariantWitnessGraph witness2 = witnesses.get(t.right);
      VariantWitnessGraphRanking ranking2 = rankings.get(t.right);
      match(witness1, witness2, ranking1, ranking2, allPotentialMatches, vertexToMatch);

      Match endMatch = getEndMatch(witness1, ranking1, witness2, ranking2);
      allPotentialMatches.add(endMatch);
    }

    return allPotentialMatches;
  }

  private Match getEndMatch(VariantWitnessGraph witness1, VariantWitnessGraphRanking ranking1, VariantWitnessGraph witness2, VariantWitnessGraphRanking ranking2) {
    TokenVertex endTokenVertex1 = witness1.getEndTokenVertex();
    TokenVertex endTokenVertex2 = witness2.getEndTokenVertex();
    Match endMatch = new Match(endTokenVertex1, endTokenVertex2);
    String sigil1 = witness1.getSigil();
    Integer rank1 = ranking1.apply(endTokenVertex1);
    endMatch.setRank(sigil1, rank1);
    String sigil2 = witness2.getSigil();
    Integer rank2 = ranking2.apply(endTokenVertex2);
    endMatch.setRank(sigil2, rank2);
    return endMatch;
  }

  List<Tuple<Integer>> permute(int max) {
    List<Tuple<Integer>> list = new ArrayList<>();
    for (int left = 0; left < max; left++) {
      for (int right = left + 1; right < max; right++) {
        list.add(new Tuple<>(left, right));
      }
    }
    return list;
  }

  private void match(VariantWitnessGraph witness1, VariantWitnessGraph witness2,//
                     VariantWitnessGraphRanking ranking1, VariantWitnessGraphRanking ranking2,//
                     Set<Match> allPotentialMatches,//
                     Map<TokenVertex, Match> vertexToMatch) {
    VariantWitnessGraphTraversal traversal1 = VariantWitnessGraphTraversal.of(witness1);
    VariantWitnessGraphTraversal traversal2 = VariantWitnessGraphTraversal.of(witness2);
    String sigil1 = witness1.getSigil();
    String sigil2 = witness2.getSigil();
    stream(traversal1)//
        .filter(SimpleTokenVertex.class::isInstance)//
        .map(SimpleTokenVertex.class::cast)//
        .forEach(tv1 -> stream(traversal2)//
            .filter(SimpleTokenVertex.class::isInstance)//
            .map(SimpleTokenVertex.class::cast)//
            .forEach(tv2 -> {
              if (matcher.apply(tv1, tv2)) {
                Match match = new Match(tv1, tv2)//
                    .setRank(sigil1, ranking1.apply(tv1))//
                    .setRank(sigil2, ranking2.apply(tv2));
                allPotentialMatches.add(match);
                vertexToMatch.put(tv1, match);
                vertexToMatch.put(tv2, match);
              }
            }));
  }

  private void visualize(DecisionTreeNode decisionTreeRootNode) {
    int indent = 0;
    visualize(decisionTreeRootNode, indent);
  }

  private void visualize(DecisionTreeNode decisionTreeNode, int indent) {
    String tab = indent == 0 ? "" : StringUtils.repeat("| ", indent - 1) + "|-";
    System.out.println(tab + decisionTreeNode.getQuantumCollatedMatchList().toString());
    for (DecisionTreeNode treeNode : decisionTreeNode.getChildNodes()) {
      visualize(treeNode, indent + 1);
    }
  }


}

File: hyper-collate-core/src/main/kotlin/nl/knaw/huygens/hypercollate/collator/HyperCollator.kt
package nl.knaw.huygens.hypercollate.collator

/*-
 * #%L
 * hyper-collate-core
 * =======
 * Copyright (C) 2017 - 2020 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import eu.interedition.collatex.dekker.Tuple
import nl.knaw.huygens.hypercollate.model.*
import nl.knaw.huygens.hypercollate.tools.CollationGraphRanking
import nl.knaw.huygens.hypercollate.tools.CollationGraphVisualizer
import org.slf4j.LoggerFactory
import java.util.*
import java.util.stream.Collectors.toSet

typealias BranchPath = List<Int>?

class HyperCollator {

    fun collate(vararg graphs: VariantWitnessGraph): CollationGraph {
        val witnesses: MutableList<VariantWitnessGraph> = graphs.sortedBy { it.sigil }.toMutableList()
        val sigils: List<String> = witnesses.map { it.sigil }
        val rankings: List<VariantWitnessGraphRanking> = witnesses.map { VariantWitnessGraphRanking.of(it) }
        val matches = getPotentialMatches(witnesses, rankings)
        val collationGraph = CollationGraph()
        val collatedTokenVertexMap: MutableMap<TokenVertex, TextNode> = HashMap()
        val first = witnesses.removeAt(0)
        val matchesSortedByRankPerWitness = sortAndFilterMatchesByWitness(matches, sigils)
        val markupNodeIndex: MutableMap<Markup, MarkupNode> = HashMap()
        initialize(collationGraph, collatedTokenVertexMap, markupNodeIndex, first)
//        visualize(collationGraph)
        for (witnessGraph in witnesses) {
            val sortedMatchesForWitness = matchesSortedByRankPerWitness[witnessGraph.sigil] ?: error("sigil not found")
            collationGraph.collate(
                    witnessGraph,
                    sortedMatchesForWitness,
                    markupNodeIndex,
                    collatedTokenVertexMap)
            collationGraph.visualize()
        }
        return collationGraph
    }

    private fun CollationGraph.visualize() {
        val dot = CollationGraphVisualizer.toDot(this, emphasizeWhitespace = true, hideMarkup = false)
        log.debug("dot={}", dot)
    }

    fun initialize(
            collationGraph: CollationGraph,
            collatedTokenVertexMap: MutableMap<TokenVertex, TextNode>,
            markupNodeIndex: MutableMap<Markup, MarkupNode>,
            witnessGraph: VariantWitnessGraph
    ) {
        val sigil = witnessGraph.sigil
        collationGraph.sigils += sigil
        markupNodeIndex.addMarkupNodes(collationGraph, witnessGraph)
        collatedTokenVertexMap[witnessGraph.startTokenVertex] = collationGraph.textStartNode
        witnessGraph.vertices()
                .asSequence()
                .filterIsInstance<SimpleTokenVertex>()
                .forEach { tokenVertex: TokenVertex ->
                    collationGraph.addCollationNode(
                            collatedTokenVertexMap,
                            tokenVertex,
                            witnessGraph,
                            markupNodeIndex)
                }
        collatedTokenVertexMap[witnessGraph.endTokenVertex] = collationGraph.textEndNode
        collationGraph.addEdges(collatedTokenVertexMap)
    }

    private fun MutableMap<Markup, MarkupNode>.addMarkupNodes(
            collationGraph: CollationGraph,
            witnessGraph: VariantWitnessGraph
    ) {
        witnessGraph
                .markupStream
                .forEach { markup: Markup ->
                    this[markup] = collationGraph.addMarkupNode(witnessGraph.sigil, markup)
                }
    }

    private fun Match.toCollatedMatch(
            sigil: String,
            collatedTokenVertexMap: Map<TokenVertex, TextNode>
    ): CollatedMatch {
        val vertex = tokenVertexList.firstOrNull { tv: TokenVertex -> tv.sigil != sigil } ?: error("No vertex!")
        val tokenVertexForWitness = getTokenVertexForWitness(sigil)
                ?: error("no TokenVertex found for sigil $sigil")
        val node = collatedTokenVertexMap[vertex] ?: error("no node found for vertex")
        return CollatedMatch(node, tokenVertexForWitness)
                .withVertexRank(getRankForWitness(sigil)!!)
    }

    private fun CollationGraph.collate(
            witnessGraph: VariantWitnessGraph,
            sortedMatchesForWitness: List<Match>,
            markupNodeIndex: MutableMap<Markup, MarkupNode>,
            collatedTokenVertexMap: MutableMap<TokenVertex, TextNode>
    ) {
        val baseRanking = CollationGraphRanking.of(this)
        val filteredSortedMatchesForWitness = sortedMatchesForWitness
                .filter { m: Match -> sigils.any { m.hasWitness(it) } }
        val witnessSigil = witnessGraph.sigil
        sigils += witnessSigil
        markupNodeIndex.addMarkupNodes(this, witnessGraph)
        val matchList = filteredSortedMatchesForWitness
                .map { it.toCollatedMatch(witnessSigil, collatedTokenVertexMap) }
                .map { m: CollatedMatch -> m.adjustRankForCollatedNode(baseRanking) }
                .distinct()
        log.debug("matchList={}, size={}", matchList, matchList.size)
        val optimalMatchList = matchList.optimized(sigils).toMutableList()
        log.debug("optimalMatchList={}, size={}", optimalMatchList, optimalMatchList.size)
        val witnessIterator: Iterator<TokenVertex> = VariantWitnessGraphTraversal.of(witnessGraph).iterator()
        val first = witnessIterator.next()
        val rootNode: TextNode = textStartNode
        collatedTokenVertexMap[first] = rootNode
        while (optimalMatchList.isNotEmpty()) {
            val match: CollatedMatch = optimalMatchList.removeAt(0)
            log.debug("match={}", match)
            val tokenVertexForWitnessGraph = match.witnessVertex
            witnessIterator.advanceWitness(
                    this,
                    collatedTokenVertexMap,
                    tokenVertexForWitnessGraph,
                    witnessGraph,
                    markupNodeIndex)
            val matchingNode = match.collatedNode
            val token = tokenVertexForWitnessGraph.token
            if (token != null) {
                matchingNode.addToken(token)
                matchingNode.addBranchPath(
                        tokenVertexForWitnessGraph.sigil, tokenVertexForWitnessGraph.branchPath)
            }
            collatedTokenVertexMap[tokenVertexForWitnessGraph] = matchingNode
            addMarkupHyperEdges(
                    witnessGraph, markupNodeIndex, tokenVertexForWitnessGraph, matchingNode)
        }
        collatedTokenVertexMap[witnessGraph.endTokenVertex] = textEndNode
        addEdges(collatedTokenVertexMap)
    }

    private fun CollationGraph.addMarkupHyperEdges(
            witnessGraph: VariantWitnessGraph,
            markupNodeIndex: Map<Markup, MarkupNode>,
            tokenVertexForWitnessGraph: TokenVertex,
            matchingNode: TextNode
    ) {
        witnessGraph
                .markupListForTokenVertex(tokenVertexForWitnessGraph)
                .forEach { markup: Markup -> linkMarkupToText(markupNodeIndex[markup] ?: error(""), matchingNode) }
    }

    private fun List<CollatedMatch>.optimized(sigils: List<String>): List<CollatedMatch> =
            OptimalCollatedMatchListAlgorithm().getOptimalCollatedMatchList(this, sigils)

    private fun CollatedMatch.adjustRankForCollatedNode(
            baseRanking: CollationGraphRanking
    ): CollatedMatch {
        nodeRank = baseRanking.apply(collatedNode)
        return this
    }

    private fun Map<TokenVertex, TextNode>.logCollated() {
        val lines: List<String> = map { (k: TokenVertex, v: TextNode) -> "${k.sigil}:${k.token} -> $v" }
        log.debug("collated={}", lines.sorted().joinToString("\n"))
    }

    private fun Iterator<TokenVertex>.advanceWitness(
            collationGraph: CollationGraph,
            collatedTokenVertexMap: MutableMap<TokenVertex, TextNode>,
            tokenVertexForWitness: TokenVertex,
            witnessGraph: VariantWitnessGraph,
            markupNodeIndex: Map<Markup, MarkupNode>
    ) {
        if (hasNext()) {
            var nextWitnessVertex = next()
            while (hasNext() && nextWitnessVertex != tokenVertexForWitness) {
                collationGraph.addCollationNode(
                        collatedTokenVertexMap,
                        nextWitnessVertex,
                        witnessGraph,
                        markupNodeIndex)
                nextWitnessVertex = next()
            }
        }
    }

    private fun CollationGraph.addCollationNode(
            collatedTokenVertexMap: MutableMap<TokenVertex, TextNode>,
            tokenVertex: TokenVertex,
            witnessGraph: VariantWitnessGraph,
            markupNodeIndex: Map<Markup, MarkupNode>
    ) {
        if (!collatedTokenVertexMap.containsKey(tokenVertex)) {
            val collationNode = addTextNodeWithTokens(tokenVertex.token!!)
            collationNode.addBranchPath(tokenVertex.sigil, tokenVertex.branchPath)
            collatedTokenVertexMap[tokenVertex] = collationNode
            addMarkupHyperEdges(witnessGraph, markupNodeIndex, tokenVertex, collationNode)
        }
    }

    fun sortAndFilterMatchesByWitness(matches: Set<Match>, sigils: List<String>): Map<String, List<Match>> =
            sigils.map { it to matches.sortedAndFilteredForWitness(it) }.toMap()

    private fun Set<Match>.sortedAndFilteredForWitness(sigil: String): List<Match> =
            filter { m: Match -> m.hasWitness(sigil) }
                    .sortedWith(sigil.matchComparator())

    private fun String.matchComparator(): Comparator<Match> =
            Comparator { match1: Match, match2: Match ->
                var rank1 = match1.getRankForWitness(this) ?: error("invalid sigil $this")
                var rank2 = match2.getRankForWitness(this) ?: error("invalid sigil $this")
                if (rank1 == rank2) {
                    rank1 = match1.getLowestRankForWitnessesOtherThan(this)
                    rank2 = match2.getLowestRankForWitnessesOtherThan(this)
                }
                rank1.compareTo(rank2)
            }

    fun getPotentialMatches(
            witnesses: List<VariantWitnessGraph>,
            rankings: List<VariantWitnessGraphRanking>
    ): Set<Match> {
        val allPotentialMatches: MutableSet<Match> = HashSet()
        val vertexToMatch: MutableMap<TokenVertex, Match> = HashMap()
        for (t in permute(witnesses.size)) {
            val witness1 = witnesses[t.left!!]
            val ranking1 = rankings[t.left!!]
            val witness2 = witnesses[t.right!!]
            val ranking2 = rankings[t.right!!]
            allPotentialMatches.addMatchesFor(witness1, witness2, ranking1, ranking2, vertexToMatch)

            val endMatch = getEndMatch(witness1, ranking1, witness2, ranking2)
            allPotentialMatches += endMatch
        }
        return allPotentialMatches
    }

    private fun getEndMatch(
            witness1: VariantWitnessGraph,
            ranking1: VariantWitnessGraphRanking,
            witness2: VariantWitnessGraph,
            ranking2: VariantWitnessGraphRanking
    ): Match {
        val endTokenVertex1 = witness1.endTokenVertex
        val endTokenVertex2 = witness2.endTokenVertex
        val endMatch = Match(endTokenVertex1, endTokenVertex2)
        val sigil1 = witness1.sigil
        val rank1 = ranking1.apply(endTokenVertex1)
        endMatch.setRank(sigil1, rank1)
        val sigil2 = witness2.sigil
        val rank2 = ranking2.apply(endTokenVertex2)
        endMatch.setRank(sigil2, rank2)
        return endMatch
    }

    private fun MutableSet<Match>.addMatchesFor(
            witness1: VariantWitnessGraph,
            witness2: VariantWitnessGraph,
            ranking1: VariantWitnessGraphRanking,
            ranking2: VariantWitnessGraphRanking,
            vertexToMatch: MutableMap<TokenVertex, Match>
    ) {
        val traversal1 = VariantWitnessGraphTraversal.of(witness1)
        val traversal2 = VariantWitnessGraphTraversal.of(witness2)
        val sigil1 = witness1.sigil
        val sigil2 = witness2.sigil
        traversal1
                .filterIsInstance<SimpleTokenVertex>()
                .forEach { tv1: SimpleTokenVertex ->
                    traversal2
                            .filterIsInstance<SimpleTokenVertex>()
                            .forEach { tv2: SimpleTokenVertex ->
                                if (tv1.matches(tv2)) {
                                    val match = Match(tv1, tv2)
                                            .setRank(sigil1, ranking1.apply(tv1))
                                            .setRank(sigil2, ranking2.apply(tv2))
                                    this += match
                                    vertexToMatch[tv1] = match
                                    vertexToMatch[tv2] = match
                                }
                            }
                }
    }

    companion object {
        private val log = LoggerFactory.getLogger(HyperCollator::class.java)

        private fun SimpleTokenVertex.matches(other: SimpleTokenVertex): Boolean {
            if (normalizedContent.isEmpty() && other.normalizedContent.isEmpty()) {
                return if (content.isEmpty() && other.content.isEmpty()) {
                    // both are milestones, so compare their tags.
                    val parentTag1 = parentXPath.replace("/.*/", "")
                    val parentTag2 = other.parentXPath.replace("/.*/", "")
                    parentTag1 == parentTag2
                } else {
                    content == other.content
                }
            }
            return normalizedContent == other.normalizedContent
        }

        private fun CollationGraph.addEdges(collatedTokenVertexMap: Map<TokenVertex, TextNode>) {
            collatedTokenVertexMap
                    .keys
                    .forEach { tv: TokenVertex ->
                        tv.incomingTokenVertexList
                                .forEach { itv: TokenVertex ->
                                    val source = collatedTokenVertexMap[itv] ?: error("source is null!")
                                    val target = collatedTokenVertexMap[tv] ?: error("target is null!")
                                    val existingTargetNodes = getOutgoingTextEdgeStream(source)
                                            .map { edge: TextEdge -> getTarget(edge) }
                                            .map { it as TextNode }
                                            .collect(toSet())
                                    val sigil = tv.sigil
                                    if (target !in existingTargetNodes) {
                                        val sigils: MutableSet<String> = mutableSetOf(sigil)
                                        addDirectedEdge(source, target, sigils)
                                        // System.out.println("> " + source + " -> " + target);
                                    } else {
                                        val edge = getOutgoingTextEdgeStream(source)
                                                .filter { e: TextEdge -> target == getTarget(e) }
                                                .findFirst()
                                                .orElseThrow { RuntimeException("There should be an edge!") }
                                        edge.addSigil(sigil)
                                    }
                                }
                    }
        }

        fun permute(max: Int): List<Tuple<Int>> {
            val list: MutableList<Tuple<Int>> = ArrayList()
            for (left in 0 until max) {
                for (right in left + 1 until max) {
                    list += Tuple(left, right)
                }
            }
            return list
        }
    }
}
